### ISSUE TITLE
Nested interceptors in Edward2

### ISSUE THREAD

**Issue by davmre** [tensorflow/probability#105]
Edward2 constructs a stack of interceptors, but program transformations defined by interceptors don't actually compose. In particular, this means it's not possible to take the log joint of a transformed model.

```python
import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import edward2 as ed

def trivial_interceptor(fn, *args, **kwargs):
   # An interceptor that does nothing.
   return fn(*args, **kwargs)

def model():
  return ed.Normal(0., 1., name="x")

def transformed_model():
  with ed.interception(trivial_interceptor):
    model()

log_joint = ed.make_log_joint_fn(model)
log_joint_transformed = ed.make_log_joint_fn(transformed_model)

print(log_joint(x=5.), log_joint_transformed(x=5.))
# expected result: identical Tensors with constant value of stdnormal.logpdf(5.)

# actual result: the log joint of the transformed model returns int(0), 
# because the interceptor inside of `make_log_joint_fn` is never invoked.
```

**Comment by dustinvtran** [comment]
+@mgorinova

Thanks for reporting this bug @davmre! It looks like the following logic happens:

1. make_log_joint's interceptor is pushed onto the stack
2. trivial_interceptor is pushed onto the stack
3. model() executes
4. ed.Normal uses the most recent interceptor, which is trivial_interceptor
5. model() terminates
6. trivial_interceptor is popped
7. make_log_joint's interceptor is popped
8. make_log_joint returns `sum([])` which is 0.

This is actually the intended behavior when I first thought about how nesting would work. However, as you point out, it's not intuitive for this example. Maybe the better behavior is that instead of using the most recent interceptor, it uses all interceptors (starting from the most recent to the oldest).

What do you think?

**Comment by mgorinova** [comment]
Thanks both for explaining what happens (and what's expected to happen) in this example! This is really helpful in terms of understanding interceptors. 

It seems to me that the intuition here breaks because interceptors return concrete values, instead of code that can be intercepted again. After `model` is intercepted by `trivial_interceptor`, it returns a random variable object (instead of an interceptable random variable constructor), so `make_log_joint`'s interceptor does nothing. (Is this sort of what happens, or am I confused?) 

This seems very related to what @davmre mentioned about algebraic effects and handlers a couple of days ago. My understanding is that if we think of interceptors as effect handlers, the correct behaviour should be to apply both interceptors (firstly `trivial_interceptor`, then `make_log_joint`'s interceptor) and get the expected result as stated by @davmre above.

**Comment by dustinvtran** [comment]
I think you're right intepreting them as effect handlers. The interceptors should pass to the next, and the oldest interceptor does actual function application (it always exists on the stack).

Change is welcome!

**Comment by mgorinova** [comment]
This has been resolved in 06f4e0023c95c08d24c2a445b6d186252869bdfd.

**Comment by mgorinova** [comment]
This has been resolved in 06f4e0023c95c08d24c2a445b6d186252869bdfd.
